### PR TITLE
ddl: fix expression index with insert causes losing index data

### LINKED ISSUE #26213: index data may be lost when creating expression index
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
``` SQL
CREATE TABLE `t` (
  `a` int(11) DEFAULT NULL,
  `b` int(11) DEFAULT NULL
);

insert into t many rows (about 1 000 000)
/* DDL */ alter table t add index id2xx((b+1));
                                    /* txn during DDL */ begin pessimistic;
                                    /* txn during DDL */ insert into t values (52000003, 503);
                                    /* txn  during DDL */  commit;
/* DDL */ finish DDL
mysql> admin check table t;
ERROR 8134 (HY000): col _V$_id2xx_0, handle 686873, index:types.Datum{k:0x0, decimal:0x0, length:0x0, i:0, collation:"", b:[]uint8(nil), x:interface {}(nil)} != record:types.Datum{k:0x1, decimal:0x0, length:0x0, i:504, collation:"", b:[]uint8(nil), x:interface {}(nil)}, compare err:<nil>
```

### 2. What did you expect to see? (Required)
No error
### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

### What problem does this PR solve?

Issue Number: close #26213 

Problem Summary:
In reorg, we can't get the value of the hidden virtual column, then the index values are wrong.

### What is changed and how it works?

For the hidden virtual column, the state is changed from none to public directly.

### Check List 

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

- No release note

### CODE REVIEW

File: ddl/index.go [hunk 1]
```diff
@@ -495,7 +498,7 @@ func (w *worker) onCreateIndex(d *ddlCtx, t *meta.Meta, job *model.Job, isPK boo
 	case model.StateNone:
 		// none -> delete only
 		indexInfo.State = model.StateDeleteOnly
-		updateHiddenColumns(tblInfo, indexInfo, model.StateDeleteOnly)
+		updateHiddenColumns(tblInfo, indexInfo, model.StatePublic)
```
Comment (AilinKid) [inline ddl/index.go:501]:
```
mysql> create table t(a int default 0, b int default 0, _V$_idx_0 int default 1);
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t(a, b) values (1, 1), (2, 2), (3, 3), (4, 4);
Query OK, 4 rows affected (0.01 sec)
Records: 4  Duplicates: 0  Warnings: 0

mysql> alter table t add index idx((b+1));
ERROR 1060 (42S21): Duplicate column name '_V$_idx_0'
```

master still has this problem, maybe we can ajust the expression index column dynamically.

**Review by AilinKid — COMMENTED** [review]
Rest LGTM

**Review by AilinKid — APPROVED** [review]

File: ddl/index.go (hunk 1 above)
Comment (wjhuang2016) [reply ddl/index.go:501]: It's exptected

File: ddl/db_test.go [hunk 2]
```diff
@@ -410,22 +410,46 @@ func (s *testSerialDBSuite) TestAddExpressionIndexRollback(c *C) {
 	ctx.Store = s.store
 	times := 0
 	hook.OnJobUpdatedExported = func(job *model.Job) {
-		if job.SchemaState == model.StateDeleteOnly {
+		switch job.SchemaState {
+		case model.StateDeleteOnly:
+			if job.SchemaState == model.StateDeleteOnly {
```
Comment (zimulala) [inline ddl/db_test.go:415]: This `if` statement is duplicated with line414.

File: ddl/db_test.go [hunk 3]
```diff
@@ -410,22 +410,46 @@ func (s *testSerialDBSuite) TestAddExpressionIndexRollback(c *C) {
...
+				if checkErr != nil {
+					return
+				}
+				_, checkErr = tk1.Exec("delete from t1 where c1 = 40;")
```
Comment (zimulala) [inline ddl/db_test.go:419]: Could we add insert and update statements here? This schema state will be running two times.

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**